internal/statesync/statesync.go

```diff
@@ -0,0 +1,316 @@
+		if err != nil {
+			ss.resetStateSync()
+			return false, false, ErrRejectSnapshot
+		}
+		ss.log.Info("DB restored")
+
+		ss.inProgress = false
+		ss.chunks = nil
+		ss.rcvdChunks = 0
+		ss.snapshot = nil
+		return true, false, nil
+	}
+
+	return false, false, nil
+}
+
+// RestoreDB restores the database from the logical sql dump using psql command
+// It also validates the snapshot hash, before restoring the database
+func RestoreDB(ctx context.Context, snapshot io.Reader,
+	dbName, dbUser, dbPass, dbHost, dbPort string,
+	snapshotHash []byte, logger log.Logger) error {
+	// unzip and stream the sql dump to psql
+	cmd := exec.CommandContext(ctx,
+		"psql",
+		"--username", dbUser,
+		"--host", dbHost,
+		"--port", dbPort,
+		"--dbname", dbName,
+		"--no-password",
+	)
+	if dbPass != "" {
+		cmd.Env = append(os.Environ(), "PGPASSWORD="+dbPass)
+	}
+
+	// cmd.Stdout = &stderr
+	stdinPipe, err := cmd.StdinPipe() // stdin for psql command
+	if err != nil {
+		return err
+	}
+	defer stdinPipe.Close()
+
+	logger.Info("Restore DB: ", log.String("command", cmd.String()))
+
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+
+	// decompress the chunk streams and stream the sql dump to psql stdinPipe
+	if err := decompressAndValidateSnapshotHash(stdinPipe, snapshot, snapshotHash); err != nil {
+		return err
+	}
+
+	stdinPipe.Close() // signifies the end of the input stream to the psql command
+
+	if err := cmd.Wait(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// decompressAndValidateChunkStreams decompresses the chunk streams and validates the snapshot hash
+func decompressAndValidateSnapshotHash(output io.Writer, reader io.Reader, snapshotHash []byte) error {
+	hasher := sha256.New()
+	_, err := io.Copy(io.MultiWriter(output, hasher), reader)
+	if err != nil {
+		return fmt.Errorf("failed to decompress chunk streams: %w", err)
+	}
+	hash := hasher.Sum(nil)
+
+	// Validate the hash of the decompressed chunks
+	if !bytes.Equal(hash, snapshotHash) {
+		return fmt.Errorf("invalid snapshot hash %x, expected %x", hash, snapshotHash)
+	}
+	return nil
+}
+
+// validateSnapshot validates the snapshot against the trusted snapshot providers
+func (ss *StateSyncer) validateSnapshot(snapshot Snapshot) error {
+	// Check if the snapshot's contents are valid
+	if snapshot.Format != DefaultSnapshotFormat {
+		return ErrUnsupportedSnapshotFormat
+	}
+
+	if snapshot.Height <= 0 || snapshot.ChunkCount <= 0 ||
+		snapshot.ChunkCount != uint32(len(snapshot.ChunkHashes)) {
+		return ErrInvalidSnapshot
+	}
+
+	// Query the snapshot providers to check if the snapshot is valid
+	height := fmt.Sprintf("%d", snapshot.Height)
+	verified := false
+	for _, clt := range ss.snapshotProviders {
+		res, err := clt.ABCIQuery(context.Background(), ABCISnapshotQueryPath, []byte(height))
+		if err != nil {
+			ss.log.Info("Failed to query snapshot", log.Error(err)) // failover to next provider
+			continue
+		}
+
+		if res.Response.Value != nil {
```

Comment:
I think it would be safer with `len(res.Response.Value) > 0`.  An empty non-nil slice might get past `json.Unmarshal` and declare an invalid snapshot instead of continuing.